### ISSUE TITLE
OF 0.12.0 Release Checklist

### ISSUE THREAD

**Issue by ofTheo** [openframeworks/openFrameworks#6692]
I found this format quite helpful to organize tracking issues for releases.
Just wanted to put this here so we can start collecting any issues/PR for the next patch release.

Please add links to issues based on testing the nightly packages on different platforms.

## 0.12.0 SPECIFIC:

Addons: 
- [ ] See if @bakercp wants to take over ofxPoco now it is not used in core ( but include ofxPoco for next few releases ) 

Libraries: 
- [ ] What libraries to drop? 
- [ ] Remove boost as dependency - but keep it in the download 

All Platforms:
- [ ] FIX Random - See @artificiel's discussion here: https://github.com/openframeworks/openFrameworks/discussions/7579 
- [x] ofDir  https://github.com/openframeworks/apothecary/pull/267
- [x] Support  C++17 - #6585 ( wip PR here #6844 )
- [x] Make ofVideoPlayer::isPlaying() consistent #6527 ( without breaking things!! ) 
- [ ] Check out proposal for avoiding ofParameter notify loop for linked ofParams #6602
- [x] ofAppGLFWWindow settings.monitor [not being clamped](https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworks/app/ofAppGLFWWindow.cpp#L240-L241) bug causes crash 
- [x] Modernize examples/templates main.cpp #6720
- [x] Better examples #7076
- [x] Bug with cairo render when setting scale to 0,0 #6715
- [x] update GLFW ( see #6954 ) 
- [x] update all 3D model loading to assimp - apart from ply - and better assimp example
- [x] update assimp for better model loading ( https://github.com/openframeworks/apothecary/pull/236 )
- [x] Drop QTCreator or fix? And focus on VS Code
- [x] Add OF_NO_FMOD to enable AVEngine Sound Player on macOS and Media Engine sound player on Windows - #7507 and #7523 
 
Windows:
- [ ] Add emscripten to Windows Release - #7484
- [x] K-Lite breaks Directshow player - #6765 
- [x] Consider a MF video player using IMFMediaEngine ( [useful ref here  ](https://discourse.libcinder.org/t/ax-mediaplayer-cinder-block-for-windows-video-playback/1851)) #7477 
- [x] Add MF and Xaudio2 based sound player #7523 

VS:
- [x] Change default to VS2019 or VS2022 
- [x] Switch to VS2019  
- [x] Remove 32bit 
- [ ] Check if C++17 fixes allow for c++latest #6832
- [x] Update PG

MSYS2:
- [x] Fix PG issue with backslashes - https://github.com/openframeworks/projectGenerator/issues/265
- [x] Remove 32bit 
- [x] Update PG 

macOS:
- [x] Change template to make missing a return a hard error instead of allowing functions not return anything. 
- [x] Fmod sound crash #6708 ( Don't see this anymore - @roymacdonald could you see if this is an issue? ) 
- [x] Assimp went from 26MB to 300MB ( 0.11.0 to 0.11.1 ) fixed by: https://github.com/openframeworks/apothecary/pull/236
- [x] Remove QTKit https://github.com/openframeworks/projectGenerator/issues/264
- [x] Fullscreen issues / viewport issues on M1 #6719
 - [x] Update template to allow serial device access #6864
 - [x] 10.15 min macOS ( once C++17 PR is merged )
 - [x] Change macOS / iOS / tvOS to ARC #6889  
 - [x] Add AVFoundation based sound player - #7293

Linux/ARM:
- [x] test Rpi 3/4 with latest changes
- [ ] Try and get Rpi 4 working without X ( [see comment from](https://github.com/openframeworks/openFrameworks/issues/6692#issuecomment-1251473646) @patriciogonzalezvivo )
- [ ] [X11 Macro clash with OpenCV 4. ](https://forum.openframeworks.cc/t/ofxcv-macro-error-on-x11/26022/22) Need to do something like this, but maybe targeted to X11 ( https://github.com/openframeworks/openFrameworks/pull/6868/files ) 
- [x] Fix Gstreamer 1.20 breaking video and grabber 
- [x] ofGstVideoPlayer fixes #7063 
- [x] Gstreamer issue with 22.04 - #7516 

Arm64 / aarch64  
- [x] OpenCv linker error https://github.com/openframeworks/openFrameworks/issues/7500#issuecomment-1518022680
- [x] Check video playback 
  
iOS:
- [x] examples are broken #6723
- [ ] guiExample broken - #7458
- [ ] native UI examples broken #7457 
- [ ] update main.cpp to modern approach
- [ ] fix simulator build not working due to duplicate symbols. 

Android:
- [ ]  #6762
- [ ] Add cpp 17 when @danoli3 's PRs are merged. 
- [ ] update main.cpp to modern approach

emscripten:
- [ ] make html template a little more modern / clean looking
- [x] check out @Jonathhhan 's [branch](https://github.com/openframeworks/openFrameworks/issues/6781#issuecomment-1023467954) for emscripten fixes. Can we do small bitesized PRs here? 
- [x] update to latest emscripten version ( see https://forum.openframeworks.cc/t/the-future-of-emscripten-and-of/38360 )
- [x] fix fullscreen and AA not working #6758 ( [see fix here](https://forum.openframeworks.cc/t/the-future-of-emscripten-and-of/38360/3?u=theo) ) 
- [x] #7377 

CI:
- [x] move nightly and releases to GitHub  
ProjectGenerator:
- [x] Remove macOS binary from PG for Linux: https://github.com/openframeworks/projectGenerator/issues/209  

Apothecary:

Examples:
- [x] - remove 3D Model loader example and rename assimpExample 

## GENERAL:
- [x] update the change log.
- [ ] update the Thanks.md with scripts/dev/thanks.sh
- [x] update version defines in ofConstants.h.
- [ ] change the version on [ofSite](https://github.com/openframeworks/ofSite/blob/master/conf.py#L1245) and make packages.
- [x] update apothecary and projectGenerator submodules
- [ ] update stable to master
- [ ] Look at issue templates again. ( First take at creating a GitHub issue template #5652 )
- [ ] Unpin this checklist and https://github.com/openframeworks/openFrameworks/issues/7145

Move to 0.13.0 
- [ ] #6989
- [ ] Remove fmod as the default sound player. #167 #5490 #6622 
- [ ] Move templates to root ( and fix CI scripts / PG / etc to correspond ) 
- [ ] Better retina support ( this is a tough one - might need to get bumped ) - https://github.com/openframeworks/openFrameworks/issues/4789
- [ ] Remove fmod as the default sound player. #167 #5490 
 - [ ] Remove fmod as the default sound player. #167 #5490 ( this seems doable if we can figure out the FFT part ) 
 - [ ] Rename osx to macOS 😬

**Comment by dimitre** [comment]
suggestions: 
maybe it is time to remove some deprecated functions like loadMovie, etc.
build RtAudio compatible with macOs 10.13

**Comment by dimitre** [comment]
### Suggestions: Folders naming and location

it would be great to change some parts of structure for a major version like 0.12.0
changing osx to macos in docs, scripts and templates

and moving templates to root, not inside scripts. it is just strange that this folder sits there, together with lots of .sh folders.

**Comment by ofTheo** [comment]
Be great if people could add any open issues not listed above to this - either in the comments or by editing this issue.

**Comment by Jonathhhan** [comment]
Regarding my Emscripten branch: Not sure what is the best way to do the bitesized pull requests, but I am open for that...
Should I make another branch with less changes (only one at a time)?
Would also be great if someone can review my changes, because there are possibly some beginner mistakes (still it works fine for me so far)...
And which of the changes should be included and which not (for now)?

I tried to make a list for a better overview (hope I did not forget anything):

Done:
Fullscreen
Font antialiasing

Done, but could be optimized:
Update OF libs for current Emscripten
Update OF libs for atomics / sharedArrayBuffer (which is needed for AudioWorklet for example)
AudioWorklet
(wasm workers were just merged into the official branch, so could be usable with the official Emscripten branch soon https://github.com/emscripten-core/emscripten/pull/12833)
WebMidi
Simplification (?) of ofxEmscriptenSoundPlayer
Routing of audio nodes (for processing `ofxEmscriptenSoundPlayer` and `ofxEmscriptenVideoPlayer` sound with audioWorklets)

Not done (but I have examples for using the file system):
Simplified use of the Emscripten file system

**Comment by ofTheo** [comment]
Bitesized Pull Requests would be great!
Especially starting with stuff which is easy to review and merge. 

People can then review the PRs in a more focused way. 


Thanks!!
Theo

**Comment by Jonathhhan** [comment]
Alright, actually I think I need somebody who can help me (already asked in the forum), also I dont have a lot of time at the moment. But I will try if I have more time.

**Comment by Jonathhhan** [comment]
Implement audioWorklet from official Emscripten branch: https://github.com/emscripten-core/emscripten/pull/16449/files

**Comment by roymacdonald** [comment]
hey @Jonathhhan  sorry for not being able to help so far. I will take a look at your branch and see how I can help to make a nice and clean PR.

**Comment by Jonathhhan** [comment]
@roymacdonald no problem. And sorry if it sounded a bit grumpy, it wasnt meant like that. I was just a little overwhelmed and exhausted.... But I think it would be a good idea to implement audioWorklets from the official Emscripten branch instead. Still, my implementation could help to show how to implement it (maybe (hopefully) it doesnt need a lot of changes).

**Comment by roymacdonald** [comment]
@Jonathhhan no worries. It did not sound like that at all to me. I see that there your most up to date branches are named `bitesize-audioworklet` and `bitesize-AudioWorklet ` which one should be used?

**Comment by Jonathhhan** [comment]
@roymacdonald those two branches were my attempts to get clean up to date branches for making the bitsize changes. But I had no luck, because I made some changes to my master branch once. So this is still the branch that includes all my changes: https://github.com/Jonathhhan/openFrameworks/tree/AudioWorklet

**Comment by dimitre** [comment]
ofxSVG object renamed to ofxSvg 
 https://github.com/openframeworks/openFrameworks/issues/6866#issuecomment-1081904974

**Comment by dimitre** [comment]
Anybody else likes the idea of having https://github.com/danomatika/ofxMidi as a core addon?

**Comment by Jonathhhan** [comment]
> Anybody else likes the idea of having https://github.com/danomatika/ofxMidi as a core addon?

definitely :)

**Comment by danoli3** [comment]
Yes yes

**Comment by dimitre** [comment]
Suggestion: this PR can be added to 0.12 checklist too
https://github.com/openframeworks/openFrameworks/pull/6889

**Comment by dimitre** [comment]
hey @ofTheo I've took a look at this:
settings.monitor not being clamped

and moved functionality to outside the if, so it happens only once
https://github.com/openframeworks/openFrameworks/pull/6954

**Comment by dimitre** [comment]
hey @ofTheo I was reading this one " Change template to make missing a return a hard error instead of allowing functions not return anything."
and it is already an error in XCode and a warning in make. 
This can be changed adding 
```
PLATFORM_CFLAGS += -Werror=return-type
```
in line 93 of config.osx.default.mk
should I submit a PR?

**Comment by ofTheo** [comment]
> 

@dimitre that would be great for make! Thanks!!

**Comment by dimitre** [comment]
Nice, its here
- https://github.com/openframeworks/openFrameworks/pull/6962

**Comment by dimitre** [comment]
Hey @ofTheo can you explain better about retina support item?

**Comment by ofTheo** [comment]
@dimitre there is is more info here: #4789 

Originally I was imagining that say you had a 2x retina scale OF would render everything twice as large into a 2x frame buffer. However fonts would also need to be loaded at 2x and then things start to get tricky as if you move the window to a 1x ( non retina window ), do you reload the fonts? Or if you start on a non retina monitor and move to a retina monitor do you reload the fonts at 2x?  

There are definitely some edge cases where things get tricky. 

It might be worth trying out some different approaches and see what is the simplest and most robust approach.  

We can query the pixel coordinate scale and we can enable retina support via the plist so, the basic info is there. 

Hope that helps!
Theo

**Comment by dimitre** [comment]
Other suggestion, older API should be removed already
loadMovie -> load
loadImage -> load
saveImage -> save
initGrabber -> setup
- https://github.com/openframeworks/openFrameworks/pull/3320

**Comment by dimitre** [comment]
Just to add to list. 
it would be great to have orientation issues solved for 0.12

**Comment by StaffanMelin** [comment]
I saw the Linux issue "Try and get Rpi 4 working without X" in the list. Does this mean running oF headless and without windowing system? I just did something similar on a Rock Pi 4 and I basically just used the ofAppNoWindow feature.

**Comment by StaffanMelin** [comment]
I think this should be addressed for Linux:
https://github.com/openframeworks/openFrameworks/issues/6871

**Comment by dimitre** [comment]
Hey @StaffanMelin do you think it is an OF issue rather than GStreamer? Is Processing still shipping a previous version binary to make it work?

**Comment by StaffanMelin** [comment]
Good question @dimitre! :) I an play videos but not capture them. I could with previous versions of gstreamer. I can also run the pipeline with gst-launch-1.0, but when running the videoGrabberExample the program crashes. So I *think* it is an oF problem when gstreamer updated beyond 1.18.

As for Processing and if it bundles an old GStreamer version, I don't know. for sure. They seem to have bundled an earlier version, then removed it, then brought it back. See https://github.com/processing/processing-video/issues/195

**Comment by patriciogonzalezvivo** [comment]
For create openGL ES 2.0 context on a RPI 4 without X11/WAYLAND (where glfw don't work) I'm doing this:
- GBM > EGL helper functions https://github.com/patriciogonzalezvivo/vera/blob/main/src/window.cpp#L195-L314
- create actual GL context https://github.com/patriciogonzalezvivo/vera/blob/main/src/window.cpp#L516-L660
For those sections ignore everything under flags `DRIVER_BROADCOM` define flags (it's for the older RPIs with a broadcom chip)

**Comment by ofTheo** [comment]
Wow - thanks @patriciogonzalezvivo! 
That's super helpful. 🙏🙏🙏

**Comment by ofTheo** [comment]
@Jonathhhan @themancalledjakob  - when you have a moment could you check the nightly builds and see if things feel good on emscripten? 
We're close to doing a 0.12.0 release so wanted to have your take a look first.

**Comment by Jonathhhan** [comment]
@ofTheo I would like to (and will do so), but I am in holidays until next thurdsday. Will definitely check it the weekend after, when I am at my computer again.

**Comment by themancalledjakob** [comment]
@ofTheo yes, sure, I should be able to find some time this weekend. There were some minor things, but overall I am using it a lot with emscripten lately and it's going pretty smooth. I'll sum it up and share quick fixes that I found asap.

**Comment by ofTheo** [comment]
Great!! Thanks @Jonathhhan @themancalledjakob - that sounds amazing. 

Also going to try and style the html to look a little cleaner before we do the release.

**Comment by ofTheo** [comment]
Just a heads up I think we'll aim to release 0.12.0 next week. 
Pinging @danoli3 for the Android stuff. If it doesn't seem doable we can roll out a 0.12.1 for Android and other fixes.

**Comment by ofTheo** [comment]
Please add anything that is must fix to the list above. 
Otherwise I think we are close and we can maybe do the Android stuff for 0.12.1

**Comment by artificiel** [comment]
if i may point 2 minor and easy-to-validate PRs: #7583 #7564

**Comment by artificiel** [comment]
ofRandom: #7598

**Comment by artificiel** [comment]
allow me to also point #7600

**Comment by artificiel** [comment]
also #7601

**Comment by artificiel** [comment]
and just in case it's an easy thing to fix (i guess it's more of a PG (or template?) thing?): make "release" the default -- as discussed in 2012: https://github.com/openframeworks/openFrameworks/issues/1022 -- there seems to be more and more performance difference between debug and release (at least macOS).

**Comment by ofTheo** [comment]
Also this: 
https://github.com/openframeworks/projectGenerator/issues/377

**Comment by ofTheo** [comment]
Just a heads up that aside from a couple of very small fixes I think we're close to 0.12.0 release. 

@dimitre if we can hold off on merging big feature or refactoring PRs for now. Let's let master sit as is for a few days and just focus on testing nightlies and addressing critical platform breaking issues. 

Goal will be to post 0.12 early next week.

**Comment by dimitre** [comment]
Great! no more news for pg until release. only bugfixes if they are found.

About RPI:
aarch64 works well for RPI 3B+ in my tests here
it seems to suport every one since RPI3 and zero 2, so this effort can make it work with any PI from after 2016
Just sharing thoughts in my opinion we can drop importance in RPI 32bits release

**Comment by ofTheo** [comment]
Nice @dimitre!  

Thanks for testing. Were you using the 64bit Raspberry Pi Bullseye OS or the Debian 64 bit OS or something different?
It would be awesome if you could try the nightly aarch64 release tomorrow as some of the libs were broken and now should be fixed.

**Comment by dimitre** [comment]
Raspberry Pi Bullseye. 
I've tested some weeks ago before the changes. 
I'll be travelling soon so no RPI near for the next two weeks. 
I'll post on the forum if somebody else can do some testing.

**Comment by ofTheo** [comment]
cool! thanks @dimitre 
I think I have 2-3 devices I can rummage up and give a try.

**Comment by artificiel** [comment]
on the x64 archlinux side, fresh install on i5 with of_v20230823_linux64gcc6_release.tar.gz: the download libs scripts work, as does compile and run the 3dprimitives example.

**Comment by ofTheo** [comment]
Awesome thanks @artificiel !

**Comment by Jonathhhan** [comment]
ProjectGenerator still does not open projects with `Open in IDE`. Tested with `of_v20230823_vs_release`.
Does not matter much to me, just want to mention it.

**Comment by ofTheo** [comment]
@Jonathhhan yup - trying to track that one down. 
If I can't solve it quickly I think we can package nightlies as 0.12.0

**Comment by ofTheo** [comment]
0.12.0 RC 1 is out here:
https://github.com/openframeworks/openFrameworks/releases/tag/0.12.0-rc1 

If there are no major issues we'll convert it to full 0.12.0. 
Thank you everyone involved for all the work getting this out. 🙏

**Comment by moebiussurfing** [comment]
Hey, thanks for the update! 

Just reporting about PG:

Dragging any example folder to PG is not allowed here on Windows 11.
Also when opening PG with admin rights.

Using the "import" button it works.
"Open IDE" button now works fine too.

**Comment by ofTheo** [comment]
Thanks for checking @moebiussurfing !!
Yes, I think the dragging into PG not working is okay ( did the older 0.11.2 release PG allow it? ). 

Many thanks!
Theo

**Comment by moebiussurfing** [comment]
> 

yes, older version allowed drag a project folder to PG

**Comment by oxillo** [comment]
On MSYS2, compilation raises a lot of annoying deprecation warnings for ofSeedRandom(). 
The message should be only emitted if the function is used by the user code. Not in  the core.

**Comment by oxillo** [comment]
Warning about compilation of C files with C++ only option

```
cc1.exe: warning: command-line option '-std=gnu++17' is valid for C++/ObjC++ but not for C
```

**Comment by artificiel** [comment]
@oxillo srand-based random will be removed from core soon, but right now the deprecation warning should only be on the argument version. (we still need a single call in the new random engine to keep the seeds in sync but that will also go away).

I've just removed instances of ofSeedRandom in the examples with #7622; can you point if you have others lines for which you get errors?

**Comment by oxillo** [comment]
Only warning that is remaining in the core is about unused function in ofFileUtils.cpp

```
C:/N/openframeworks/of_v0.12.0-rc1_msys2_mingw64_release/libs/openFrameworks/utils/ofFileUtils.cpp:1480:13: warning: 'bool natural(const ofFile&, const ofFile&)' defined but not used [-Wunused-function]
```

I think it can be safely deleted as it is replace by naturalStr(). @ofTheo what do you think ?

**Comment by ofTheo** [comment]
Closing this issue! 
https://github.com/openframeworks/openFrameworks/releases/tag/0.12.0
https://forum.openframeworks.cc/t/0-12-0-is-released/42426

Thanks everyone!!!! 🎉
